src/linear_algebra/affine_space/restrict.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2022 Paul Reichert. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Reichert
+-/
+import linear_algebra.affine_space.affine_subspace
+
+/-!
+# Affine map restrictions
+
+This file defines restrictions of affine maps.
+
+## Main definitions
+
+* The domain and codomain of an affine map can be restricted using
+  `affine_map.restrict`.
+
+## Main theorems
+
+* The associated linear map of the restriction is the restriction of the
+  linear map associated to the original affine map.
+* The restriction is injective if the original map is injective.
+* The restriction in surjective if the codomain is the image of the domain.
+-/
+
+variables {k V₁ P₁ V₂ P₂ : Type*} [ring k]
+  [add_comm_group V₁] [add_comm_group V₂]
+  [module k V₁] [module k V₂]
+  [add_torsor V₁ P₁] [add_torsor V₂ P₂]
+
+include V₁ V₂
+
+/- not an instance because it loops with `nonempty` -/
+lemma affine_subspace.nonempty_map {E : affine_subspace k P₁} [Ene : nonempty E]
+  {φ : P₁ →ᵃ[k] P₂} : nonempty (E.map φ) :=
+begin
+  obtain ⟨x, hx⟩ := id Ene,
+  refine ⟨⟨φ x, affine_subspace.mem_map.mpr ⟨x, hx, rfl⟩⟩⟩,
+end
+
+local attribute [instance, nolint fails_quickly] affine_subspace.nonempty_map
+local attribute [instance, nolint fails_quickly] affine_subspace.to_add_torsor
+
+/-- Restrict domain and codomain of an affine map to the given submodules. -/
+def affine_map.restrict
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) : E →ᵃ[k] F :=
+begin
+  refine ⟨_, _, _⟩,
+  { exact λ x, ⟨φ x, hEF $ affine_subspace.mem_map.mpr ⟨x, x.property, rfl⟩⟩ },
+  { refine φ.linear.restrict (_ : E.direction ≤ F.direction.comap φ.linear),
+    rw [←submodule.map_le_iff_le_comap, ←affine_subspace.map_direction],
+    exact affine_subspace.direction_le hEF },
+  { intros p v,
+    simp only [subtype.ext_iff, subtype.coe_mk, affine_subspace.coe_vadd],
+    apply affine_map.map_vadd },
+end
+
+lemma affine_map.restrict.coe_apply
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) (x : E) :
+  ↑(φ.restrict hEF x) = φ x := rfl
+
+lemma affine_map.restrict.linear
+  (φ : P₁ →ᵃ[k] P₂) {E : affine_subspace k P₁} {F : affine_subspace k P₂}
+  [nonempty E] [nonempty F]
+  (hEF : E.map φ ≤ F) :
+  (φ.restrict hEF).linear = φ.linear.restrict
+    (by { change E.direction ≤ F.direction.comap φ.linear,
+          rw [←submodule.map_le_iff_le_comap, ←affine_subspace.map_direction],
+          exact affine_subspace.direction_le hEF }) := rfl
```

Comment:
This suggests the inline `by { change, rw, exact }` proof could be its own lemma, to help with readability.

Reply:
I'm just a bit troubled with introducing such helper lemmas because I fear that this would be overly specialized and not very easily nameable or useful in other contexts: "If the image of `E` is contained in `F`, then `x` being in `E.direction` implies its image to be in `F.direction`." The way I tried to improve this was by introducing the "nicer" `linear_map.restrict'`, which I removed from the PR again (see previous discussion).
Do you think that an extracted lemma would be of independent use?

Reply:
In this case, I would define a `lemma affine_map.restrict.linear_aux` solely for this purpose.
It is not uncommon to do so, and it is still perfectly fine if it only has a one time use.
I was hesitant about `linear_map.restrict'` because it was a declaration, but it usually doesn't hurt when adding wrappers to lemmas.

Reply:
Great, thanks!